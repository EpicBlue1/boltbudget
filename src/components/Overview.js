import React from 'react';
import {useEffect, useState} from 'react';
import TableItem from './subComps/userTabItem';
import TableItemTwo from './subComps/expTabItem';
import SavingsSec from './Savings';
import Modal from './subComps/infoModal';

const Overview = (props) => {

    const [tableItemOne, settableItemOne] = useState();
    const [tableItemTwo, settableItemTwo] = useState();
    const [counter, setcounter] = useState(0);

    const AddUserSec = (e) => {
        e.preventDefault();
        let FirstName = document.getElementById('FirstName').value;
        let LastName = document.getElementById('LastName').value;
        let Salary = parseInt(document.getElementById('Salary').value);
        let obj = {name: FirstName + " " + LastName, salary: Salary}

        let counter = 1;
        counter = counter + 1;

        setcounter(counter)
        
        props.UsersBef.push(obj);

        document.getElementById('FirstName').value = "";
        document.getElementById('LastName').value = "";
        document.getElementById('Salary').value = "";
    }

    const AddExp = (e) => {
        e.preventDefault();
        let ProductName = document.getElementById('ProName').value;
        let Cost = parseInt((document.getElementById('Cost').value) + ".00");
        let obj = {prod_name: ProductName, cost: Cost} 
        
        let counter = 1;
        counter = counter + 1;

        setcounter(counter)
        
        props.expenses.push(obj);
        console.log(props.expenses)
    }

    useEffect(() => {
        console.log(props.totTaxes);
        let renderTb = props.totTaxes.map((item) => <TableItem name={item.name} salary={item.salary} salaryAfterTax={item.salary - item.salaryAfterTax/12} taxBracket={item.taxB}/>);
        settableItemOne(renderTb);

        let renderTbTwo = props.expenses.map((item) => <TableItemTwo name={item.prod_name} price={item.cost}/>);
        settableItemTwo(renderTbTwo);
    });

    return (
        <div id="section1" className="Container">
            {/* <Modal /> */}
            <h1  className="Title">
                Overview
            </h1>
            <div className="OverviewSub">
                <h2 className="HeadingTwo inline">Total Income (Household)</h2>
                <h2 className="HeadingTwo inline">Total Expenses</h2>
                <h2 className="HeadingTwo inline">Total Tax Expense (Household)</h2>
                <h2 className="HeadingTwo inline">R{props.totalIncome}</h2>
                <h2 className="HeadingTwo inline">R{props.totExp}</h2>
                <h2 className="HeadingTwo inline">R{props.totAfterTax}</h2>
            </div>
            <hr></hr>
            <h2 className="TitleTwo">
                What is Bolt Budget?
            </h2>
            <div className="OverviewSub TitleTwo">
                <p>lorem ipsum dolor</p>
            </div>
            <br></br>
            <hr></hr>
            <h1 className="Title">
                Users
            </h1>
            <div id="section1" className="AddUserSec">
            <h2 className="">Add a user</h2>
                <form onSubmit={AddUserSec}>
                    <input data-testid="Fname" id='FirstName' required placeholder="First Name" className="addUser"/>
                    <input data-testid="Lname" id='LastName' required placeholder="Last Name" className="addUser"/>
                    <input data-testid="Income" id='Salary' required type="number" placeholder="Income" className="addUser"/>
                    <br></br>
                    <br></br>
                    <button data-testid="AddUserBut" className="addUser">Add User</button>
                </form>
            </div>
            <div data-testid="UsersTable" className="UserSec">
                <br></br>
                <table>
                <tr className="TBtitles">
                    <th>Name</th>
                    <th>Income</th>
                    <th>Tax taxBracket</th>
                    <th>Income After Tax</th>
                </tr>
                    {tableItemOne}
                </table>
            </div>
            <hr></hr>
            <h1 className="Title">
                Expenses
            </h1>
            <div id="section2" className="AddUserSec">
            <h2 className="">Add an expense</h2>
                <form onSubmit={AddExp}>
                    <input data-testid="ProductName" id='ProName' required placeholder="Product Name" className="addUser"/>
                    <input data-testid="ProCost" id='Cost' required type="number" placeholder="Cost" className="addUser"/>
                    <br></br>
                    <br></br>
                    <button data-testid="AddExpense" className="addUser">Add Expense</button>
                </form>
            </div>
            <div className="UserSec">
                <br></br>
                <table>
                <tr className="TBtitles">
                    <th>Product</th>
                    <th>Price</th>
                </tr>
                    {tableItemTwo}
                </table>
            </div>
            <hr></hr>
            <h1 className="Title">
                Savings
            </h1>
            <SavingsSec totTaxes={props.totTaxes} totAfterTax={props.totAfterTax} totalIncome={props.totalIncome}/>
        </div>
    );
};

export default Overview;