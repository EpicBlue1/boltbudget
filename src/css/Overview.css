.AddUserSec{
    float:left;
    width: 30%;
    height: 26vh;
  }

  .UserSec{
    float:left;
    width: 70%;
    height: 26vh;
  }

  .addUser{
    margin-right: 1%;
  }


