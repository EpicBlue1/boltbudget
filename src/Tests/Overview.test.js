import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Overview from '../components/Overview';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

describe('Testing Overview functions excluding savings section', () => {

    const MockUserTaxes = [{name: "Johnny Bravo", salary: 67000, salaryAfterTax: 52330.166666666664, taxB: "170 734 + 39% of taxable income above 641 400"},
                        {name: "Susan Golden", salary: 20000, salaryAfterTax: 14159.5, taxB: "40 680 + 26% of taxable income above 226 000"}]

    const MockUserExpenses = [{prod_name: 'Dairy Milk 1L', cost: 19.99}, {prod_name: 'Dairy Milk 1L', cost: 19.99}]

    const Before = ([
        {name: 'Johan Golden', salary: 67000}, 
        {name: 'Susan Golden', salary: 20000}
    ])

    const userInput = ({inputOne, inputTwo, inputThree, inputFour, inputFive}) => {

        const inputFnameElement = screen.getByTestId(/Fname/i);
        const inputLnameElement = screen.getByTestId(/Lname/i);
        const inputIncomeElement = screen.getByTestId(/Income/i); 

        //expenses section
        const inputProductElement = screen.getByTestId(/ProductName/i);
        const inputCostElement = screen.getByTestId(/ProCost/i); 
        
        if(inputOne){
            userEvent.type(inputFnameElement, inputOne)
        }
        if(inputTwo){
            userEvent.type(inputLnameElement, inputTwo)
        }
        if(inputThree){
            userEvent.type(inputIncomeElement, inputThree)
        }
    //expenses section
        if(inputFour){
            userEvent.type(inputProductElement, inputFour)
        }
        if(inputFive){
            userEvent.type(inputCostElement, inputFive)
        }
        return {
            inputFnameElement,
            inputLnameElement,
            inputIncomeElement,
            inputCostElement,
            inputProductElement
        }
    }

    beforeEach(() => {
        console.log("Overview Functions Test")
        render(<Overview totTaxes={MockUserTaxes} UsersBef={Before} expenses={MockUserExpenses} />);
    })

    test('make sure add User section inputs are empty when component is rendered', () => {
        const {inputFnameElement, inputLnameElement, inputIncomeElement} = userInput({});

        userInput({})

        expect(inputFnameElement.value).toBe("");
        expect(inputLnameElement.value).toBe("");
        expect(inputIncomeElement.value).toBe("");
    })

    test('make sure add Expense section inputs are empty when component is rendered', () => {
        const {inputProductElement, inputCostElement} = userInput({});

        userInput({})

        expect(inputProductElement.value).toBe("");
        expect(inputCostElement.value).toBe("");
    })

    test("Test if mock array users (2) and data Object Array is added to dom", () =>{
        //users
        const resultElementOne = screen.getByText("Johnny", {exact: false});
        expect(resultElementOne.textContent).toBe("Johnny Bravo")

        const resultElementTwo = screen.getByText("Susan", {exact: false});
        expect(resultElementTwo.textContent).toBe("Susan Golden")

        //taxBrackets
        const TaxBracketOne = screen.getByText("170 734", {exact: false});
        expect(TaxBracketOne.textContent).toBe("170 734 + 39% of taxable income above 641 400")
        
        const TaxBracketTwo = screen.getByText("40 680", {exact: false});
        expect(TaxBracketTwo.textContent).toBe("40 680 + 26% of taxable income above 226 000")
    })

    //Expenses
    test("Test if mock array users (2) and data Object Array is added to dom", () =>{
        //Expenses
        const resultElementOne = screen.getByText("Dairy", {exact: false});
        expect(resultElementOne.textContent).toBe("Dairy Milk 1L")

        const resultElementTwo = screen.getByText("Albany", {exact: false});
        expect(resultElementTwo.textContent).toBe("Albany White Bread")

        //Price
        const TaxBracketOne = screen.getByText("R19.99", {exact: false});
        expect(TaxBracketOne.textContent).toBe("R19.99")
        
        const TaxBracketTwo = screen.getByText("R14.99", {exact: false});
        expect(TaxBracketTwo.textContent).toBe("R14.99")
    })

    it('If overview excluding savings renders correctly', () => {
        const component = renderer.create(<Overview totTaxes={MockUserTaxes} UsersBef={Before} expenses={MockUserExpenses}/>).toJSON();
    
        expect(component).toMatchSnapshot();
    });
});