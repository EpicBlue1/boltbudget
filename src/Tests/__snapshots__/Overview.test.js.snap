// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Overview functions excluding savings section If overview excluding savings renders correctly 1`] = `
<div
  className="Container"
  id="section1"
>
  <h1
    className="Title"
  >
    Overview
  </h1>
  <div
    className="OverviewSub"
  >
    <h2
      className="HeadingTwo inline"
    >
      Total Income (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Expenses
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Tax Expense (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
  </div>
  <hr />
  <h2
    className="TitleTwo"
  >
    What is Bolt Budget?
  </h2>
  <div
    className="OverviewSub TitleTwo"
  >
    <p>
      lorem ipsum dolor
    </p>
  </div>
  <br />
  <hr />
  <h1
    className="Title"
  >
    Users
  </h1>
  <div
    className="AddUserSec"
    id="section1"
  >
    <h2
      className=""
    >
      Add a user
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="Fname"
        id="FirstName"
        placeholder="First Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Lname"
        id="LastName"
        placeholder="Last Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Income"
        id="Salary"
        placeholder="Income"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddUserBut"
      >
        Add User
      </button>
    </form>
  </div>
  <div
    className="UserSec"
    data-testid="UsersTable"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Name
        </th>
        <th>
          Income
        </th>
        <th>
          Tax taxBracket
        </th>
        <th>
          Income After Tax
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Expenses
  </h1>
  <div
    className="AddUserSec"
    id="section2"
  >
    <h2
      className=""
    >
      Add an expense
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="ProductName"
        id="ProName"
        placeholder="Product Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="ProCost"
        id="Cost"
        placeholder="Cost"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddExpense"
      >
        Add Expense
      </button>
    </form>
  </div>
  <div
    className="UserSec"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Product
        </th>
        <th>
          Price
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Savings
  </h1>
  <div
    className="SavSec"
    id="section3"
  >
    <div
      className="SavingsCol float"
    >
      <p
        className="float"
      >
        Calculate after tax
      </p>
      <input
        aria-label="CheckBox"
        className="CHECK"
        onClick={[Function]}
        type="checkbox"
      />
      <br />
      <br />
      <h2
        className="SavingHeading"
      >
        Percentage
      </h2>
      <input
        data-testid="PercInput"
        id="PercVal"
        placeholder="%"
        type="numer"
      />
      <button
        data-testid="CalculatePerc"
        onClick={[Function]}
      >
        Calculate
      </button>
    </div>
    <div
      className="SavingsCol float"
    >
      <h2
        className="SavingHeading"
      >
        Potential to save
      </h2>
      <p
        data-testid="TotalToSave"
      >
        R 
      </p>
      <h2
        className="SavingHeading"
      >
        Total Saved monthly
      </h2>
      <p
        data-testid="Saved"
      >
        R 
        Type in a percentage
      </p>
    </div>
  </div>
</div>
`;

exports[`Testing Overview functions excluding savings section When Values inputs and values are updated 1`] = `
<div
  className="Container"
  id="section1"
>
  <h1
    className="Title"
  >
    Overview
  </h1>
  <div
    className="OverviewSub"
  >
    <h2
      className="HeadingTwo inline"
    >
      Total Income (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Expenses
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Tax Expense (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
  </div>
  <hr />
  <h2
    className="TitleTwo"
  >
    What is Bolt Budget?
  </h2>
  <div
    className="OverviewSub TitleTwo"
  >
    <p>
      lorem ipsum dolor
    </p>
  </div>
  <br />
  <hr />
  <h1
    className="Title"
  >
    Users
  </h1>
  <div
    className="AddUserSec"
    id="section1"
  >
    <h2
      className=""
    >
      Add a user
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="Fname"
        id="FirstName"
        placeholder="First Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Lname"
        id="LastName"
        placeholder="Last Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Income"
        id="Salary"
        placeholder="Income"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddUserBut"
      >
        Add User
      </button>
    </form>
  </div>
  <div
    className="UserSec"
    data-testid="UsersTable"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Name
        </th>
        <th>
          Income
        </th>
        <th>
          Tax taxBracket
        </th>
        <th>
          Income After Tax
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Expenses
  </h1>
  <div
    className="AddUserSec"
    id="section2"
  >
    <h2
      className=""
    >
      Add an expense
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="ProductName"
        id="ProName"
        placeholder="Product Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="ProCost"
        id="Cost"
        placeholder="Cost"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddExpense"
      >
        Add Expense
      </button>
    </form>
  </div>
  <div
    className="UserSec"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Product
        </th>
        <th>
          Price
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Savings
  </h1>
  <div
    className="SavSec"
    id="section3"
  >
    <div
      className="SavingsCol float"
    >
      <p
        className="float"
      >
        Calculate after tax
      </p>
      <input
        aria-label="CheckBox"
        className="CHECK"
        onClick={[Function]}
        type="checkbox"
      />
      <br />
      <br />
      <h2
        className="SavingHeading"
      >
        Percentage
      </h2>
      <input
        data-testid="PercInput"
        id="PercVal"
        placeholder="%"
        type="numer"
      />
      <button
        data-testid="CalculatePerc"
        onClick={[Function]}
      >
        Calculate
      </button>
    </div>
    <div
      className="SavingsCol float"
    >
      <h2
        className="SavingHeading"
      >
        Potential to save
      </h2>
      <p
        data-testid="TotalToSave"
      >
        R 
      </p>
      <h2
        className="SavingHeading"
      >
        Total Saved monthly
      </h2>
      <p
        data-testid="Saved"
      >
        R 
        Type in a percentage
      </p>
    </div>
  </div>
</div>
`;

exports[`Testing Overview functions excluding savings section renders correctly 1`] = `
<div
  className="Container"
  id="section1"
>
  <h1
    className="Title"
  >
    Overview
  </h1>
  <div
    className="OverviewSub"
  >
    <h2
      className="HeadingTwo inline"
    >
      Total Income (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Expenses
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      Total Tax Expense (Household)
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
    <h2
      className="HeadingTwo inline"
    >
      R
    </h2>
  </div>
  <hr />
  <h2
    className="TitleTwo"
  >
    What is Bolt Budget?
  </h2>
  <div
    className="OverviewSub TitleTwo"
  >
    <p>
      lorem ipsum dolor
    </p>
  </div>
  <br />
  <hr />
  <h1
    className="Title"
  >
    Users
  </h1>
  <div
    className="AddUserSec"
    id="section1"
  >
    <h2
      className=""
    >
      Add a user
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="Fname"
        id="FirstName"
        placeholder="First Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Lname"
        id="LastName"
        placeholder="Last Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="Income"
        id="Salary"
        placeholder="Income"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddUserBut"
      >
        Add User
      </button>
    </form>
  </div>
  <div
    className="UserSec"
    data-testid="UsersTable"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Name
        </th>
        <th>
          Income
        </th>
        <th>
          Tax taxBracket
        </th>
        <th>
          Income After Tax
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Expenses
  </h1>
  <div
    className="AddUserSec"
    id="section2"
  >
    <h2
      className=""
    >
      Add an expense
    </h2>
    <form
      onSubmit={[Function]}
    >
      <input
        className="addUser"
        data-testid="ProductName"
        id="ProName"
        placeholder="Product Name"
        required={true}
      />
      <input
        className="addUser"
        data-testid="ProCost"
        id="Cost"
        placeholder="Cost"
        required={true}
        type="number"
      />
      <br />
      <br />
      <button
        className="addUser"
        data-testid="AddExpense"
      >
        Add Expense
      </button>
    </form>
  </div>
  <div
    className="UserSec"
  >
    <br />
    <table>
      <tr
        className="TBtitles"
      >
        <th>
          Product
        </th>
        <th>
          Price
        </th>
      </tr>
    </table>
  </div>
  <hr />
  <h1
    className="Title"
  >
    Savings
  </h1>
  <div
    className="SavSec"
    id="section3"
  >
    <div
      className="SavingsCol float"
    >
      <p
        className="float"
      >
        Calculate after tax
      </p>
      <input
        aria-label="CheckBox"
        className="CHECK"
        onClick={[Function]}
        type="checkbox"
      />
      <br />
      <br />
      <h2
        className="SavingHeading"
      >
        Percentage
      </h2>
      <input
        data-testid="PercInput"
        id="PercVal"
        placeholder="%"
        type="numer"
      />
      <button
        data-testid="CalculatePerc"
        onClick={[Function]}
      >
        Calculate
      </button>
    </div>
    <div
      className="SavingsCol float"
    >
      <h2
        className="SavingHeading"
      >
        Potential to save
      </h2>
      <p
        data-testid="TotalToSave"
      >
        R 
      </p>
      <h2
        className="SavingHeading"
      >
        Total Saved monthly
      </h2>
      <p
        data-testid="Saved"
      >
        R 
        Type in a percentage
      </p>
    </div>
  </div>
</div>
`;
